File: 0929-unique-email-addresses/0929-unique-email-addresses.rb
# @param {String[]} emails
# @return {Integer}
def num_unique_emails(emails)
  emails.uniq { |email| clean(email) }.count
end

def clean(email)
  local, domain = email.split('@')
  local.split('+').first.split('.').join + '@' + domain
end

File: 0929-unique-email-addresses/0929-unique-email-addresses.cpp
class Solution {
public:
    int numUniqueEmails(vector<string>& emails) {
        unordered_set<string> uniq_emails;
        for (string& email: emails) {
            uniq_emails.insert(clean(email));
        }
        return uniq_emails.size();
    }
    
private:
    string clean(string& email) {
        stringstream ss(email);
        string local, domain;
        getline(ss, local, '@');
        getline(ss, domain);

        size_t plus_pos = local.find('+');
        if (plus_pos != string::npos) {
            local = local.substr(0, plus_pos);
        }
        local.erase(remove(local.begin(), local.end(), '.'), local.end());

        return local + "@" + domain;
    }
};